src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerImpl.scala

```diff
@@ -103,43 +103,54 @@ class GroupManagerImpl(
   override def pod(id: PathId): Option[PodDefinition] = rootGroup().pod(id)
 
   @SuppressWarnings(Array("all")) /* async/await */
-  override def updateRoot(
+  override def updateRootEither[T](
     id: PathId,
-    change: (RootGroup) => RootGroup, version: Timestamp, force: Boolean, toKill: Map[PathId, Seq[Instance]]): Future[DeploymentPlan] = {
+    change: (RootGroup) => Either[T, RootGroup],
+    version: Timestamp, force: Boolean, toKill: Map[PathId, Seq[Instance]]): Future[Either[T, DeploymentPlan]] = try {
 
     // All updates to the root go through the work queue.
-    val deployment = serializeUpdates {
-      async {
-        logger.info(s"Upgrade root group version:$version with force:$force")
-
-        val from = rootGroup()
-        val unversioned = assignDynamicServicePorts(from, change(from))
-        val to = GroupVersioningUtil.updateVersionInfoForChangedApps(version, from, unversioned)
-        Validation.validateOrThrow(to)(RootGroup.rootGroupValidator(config.availableFeatures))
-        val plan = DeploymentPlan(from, to, version, toKill)
-        Validation.validateOrThrow(plan)(DeploymentPlan.deploymentPlanValidator())
-        logger.info(s"Computed new deployment plan:\n$plan")
-        await(groupRepository.storeRootVersion(plan.target, plan.createdOrUpdatedApps, plan.createdOrUpdatedPods))
-        await(deploymentService.get().deploy(plan, force))
-        await(groupRepository.storeRoot(plan.target, plan.createdOrUpdatedApps, plan.deletedApps, plan.createdOrUpdatedPods, plan.deletedPods))
-        logger.info(s"Updated groups/apps/pods according to plan ${plan.id}")
-        // finally update the root under the write lock.
-        root := Option(plan.target)
-        plan
+    val maybeDeploymentPlan: Future[Either[T, DeploymentPlan]] = serializeUpdates {
+      logger.info(s"Upgrade root group version:$version with force:$force")
+
+      val from = rootGroup()
```

Comment:
I read it few times but I still don't completely understand what motivated the change here - can you maybe explain more, since I don't know this part of code that well. Before your change, it always returned a deployment plan and now it does not have to? What changed?

Reply:
### Before:
This method returned deployment plan and threw an exception if something would go wrong.
### Now:
This method returns deployment plan and an error object (in the most cases - rejection) explicitly using `Either`.
### Motivation:
In akka-http we explicitly handle **errors** (via rejections) and throw exceptions only when unexpected **failure** happened. 

Difference between these two is that **failure is an unexpected event** that prevents us from continuing to function normally. In contrast an **error is an expected event**, which we know how to handle, (for validation error or other hand-written thing). Errors will be communicated to the client as part of the normal processing of the message. 

You can find more here: https://www.reactivemanifesto.org/glossary#Failure

Reply:
I do like the approach. However, `Rejections` should not leak into this area. Ideally we'd avoid not found and not authorized errors before we even call the logic layer. I used disjunctions in previous projects but I don't think we should introduce Scalaz or Cats. What kind of expected error can we not handle early on the controller layer?

Reply:
@meln1k thanks for the explanation 👍